### ISSUE TITLE
Spatial pooler: remove tieBreaker functionality for overlapping columns

### ISSUE THREAD

**Issue by breznak** [numenta/nupic#2629]
`tieBreaker` is an array of random noise, used to "break" ties between columns with the same overlap when we need to sort them for inhibition/boosting. 

As @scottpurdy suggested in https://github.com/numenta/nupic.core/pull/629#discussion_r41069153 the array should be removed and replaced with a function that will break individual ties. 

Benefit: columns win ties randomly, noone gets a preference. 
Possible negative: speed (?) 

TODO: 
- [ ] needs appropriate change in cpp too

**Comment by scottpurdy** [comment]
@subutai @chetan51 - Do you guys see anything wrong with removing the tieBreaker array in favor of generating random tie breaker values when needed? It seems to me that we could not add any noise and just sort, letting the result of the sort be the tie breaker.

The only reason I see to keep it as is is if we need the column preference (which column wins out over which other columns in a tie breaker) to be the same over time. This doesn't seem beneficial to me though.

**Comment by subutai** [comment]
The main issue is that the tie breaking should be consistent. That way a given input will always generate the same SDR (assuming no learning in between). This is why we use a random but fixed tie breaker array.

**Comment by breznak** [comment]
> The main issue is that the tie breaking should be consistent.

shouldn't sorting (even the same values) be also deterministic? But then
the array seems safer

**Comment by subutai** [comment]
> shouldn't sorting (even the same values) be also deterministic?

Yeah, we could do that. The random array spreads out the winners a bit more but probably has identical behavior since it is a fixed random array.

**Comment by breznak** [comment]
> Yeah, we could do that. The random array spreads out the winners a bit
more but probably has identical behavior since it is a fixed random array.

Question is if it's worth the work.

Or we could replace the `tieBreaker` array from time to time, with a fixed
rng it's deterministic + has the benefit of modified "biased columns" (if
that's significant at all)

**Comment by subutai** [comment]
> Question is if it's worth the work.

I don't think it's worth the work. We would need to update both PY and C++ implementations. I doubt it will significantly improve performance or accuracy in any way.

**Comment by breznak** [comment]
Marking a P4, or close if you feel like that.

**Comment by scottpurdy** [comment]
@subutai - It would be nice to lose the serialized state. We could deterministically determine the random seeds, store them in memory for speed purposes, but regenerate them whenever we reload the model. That keeps same behavior and speed but without the extra serialized state. Thoughts on that?

**Comment by subutai** [comment]
@scottpurdy Sure, that would be fine. I hadn't considered the cost of serialization, so that seems like a win.

**Comment by cogmission** [comment]
Why does it matter if the RNG sequence repeats? It only executes random
choosing behavior, it doesn't really determine much...? Why bother
serializing it? Isn't that taking determinism too far? Who cares if you
pick the same learning cell at any particular point in the sequence? Things
like that are deterministic at a higher level where picking cell 1 or cell
2 will result in the same distal dendrite architecture - it will all just
connect to a different cell, but learning and recognized sequences will all
be the same? (when using the TM as an example, but the same holds true for
the places the RNG is used in the SP).

The poignant level at which determinism is important is the inferential
results not which specific cell is responsible...? Just a thought...

On Wed, Oct 7, 2015 at 1:08 PM, Subutai Ahmad <notifications@github.com>
wrote:

> @scottpurdy <https://github.com/scottpurdy> Sure, that would be fine. I
> hadn't considered the cost of serialization, so that seems like a win.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/2629#issuecomment-146280775>.
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*Cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io
